### ISSUE TITLE
keep tags in memory

### ISSUE THREAD

**Issue by rainerborene** [neovim/neovim#1264]
I've noticed that Vim could be really slow when you've lots of tags files. I'm using tpope's vim-bundler plugin which appends each gem directory to your `tags` and `path` options. It would be a good idea to cache these files in memory? https://gist.github.com/rainerborene/a0a504fc98b8d18ec98b

Looking at the source code the `find_tags()` function seems to be the problem. The mechanism by which vim finds a tag and jumps to it is, roughly, this:

```
foreach tagfile in &tags
    search for $tag in $tagfile
    if found:
        get filename from $tagfile
        foreach dir in &path:
             if $filename in $dir:
                 open $filename to $tag
```

With that in mind, there are 2 points at which I think could be hitting a delay:

```
Opening $tagname
Checking for $filename in $dir
```

Same thing happen when you use `C-X C-]` keystrokes for auto completion.

**Comment by justinmk** [comment]
@jamessan Does https://github.com/vim/vim/commit/810f9c361c83afb36b9f1cdadca2b93f1201d039 help with this?

**Comment by jamessan** [comment]
> @jamessan Does vim/vim@810f9c3 help with this?

Yes, I've started merging that and the related follow up patches.  I'll close this with that PR, since the performance improvements there should be more than enough.

**Comment by jamessan** [comment]
The tag lookup improvements merged in df1e7b7 fix the performance problems.

**Comment by jamessan** [comment]
For comparison, here's a rough before/after using a tags file generated by `ctags -R /usr/include`.

* Pre-merge:
  
      $ time ./build/bin/nvim -u NONE --cmd 'echo len(taglist("a"))."\n"' --cmd q
      135603
      ./build/bin/nvim -u NONE --cmd 'echo len(taglist("a"))."\n"' --cmd q  85.02s user 2.52s system 99% cpu 1:27.70 total

* Post-merge:
  
      $ time ./build/bin/nvim -u NONE --cmd 'echo len(taglist("a"))."\n"' --cmd q
      135603
      ./build/bin/nvim -u NONE --cmd 'echo len(taglist("a"))."\n"' --cmd q  0.50s user 0.06s system 99% cpu 0.568 total
